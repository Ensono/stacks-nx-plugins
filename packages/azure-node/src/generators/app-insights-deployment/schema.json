{
  "$schema": "http://json-schema.org/schema",
  "$id": "AppInsightsDeployment",
  "title": "",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "alias": "p",
      "$default": {
        "$source": "projectName"
      },
      "x-prompt": "What is the name of the project for this component?"
    },
    "applicationinsightsConnectionString": {
      "type": "string",
      "description": "The env variable that stores the app insights connection string.",
      "alias": "e",
      "default": "APPLICATIONINSIGHTS_CONNECTION_STRING",
      "x-prompt": "What is the env variable for the application insights connection string?",
      "pattern": "^[A-Z]+(?:_[A-Z]+)*$",
      "minLength": 5,
      "maxLength": 100
    }
  },
  "required": [
    "project",
    "applicationinsightsConnectionString"
  ]
}
