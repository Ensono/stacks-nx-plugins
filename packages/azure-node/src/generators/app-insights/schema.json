{
    "$schema": "http://json-schema.org/schema",
    "cli": "nx",
    "$id": "AppInsights",
    "title": "",
    "type": "object",
    "properties": {
        "project": {
            "type": "string",
            "description": "The name of the project.",
            "alias": "p",
            "$default": {
                "$source": "projectName"
            },
            "x-prompt": "What is the name of the project for this component?"
        },
        "appInsightsKey": {
            "type": "string",
            "description": "The env variable that stores the app insights key.",
            "alias": "e",
            "x-prompt": "What is your application insights key?"
        },
        "server": {
            "type": "string",
            "description": "The name of your custom server file",
            "alias": "s",
            "x-prompt": "What is the path within your app to the server?"
        }
    },
    "required": ["project", "appInsightsKey", "server"]
}
