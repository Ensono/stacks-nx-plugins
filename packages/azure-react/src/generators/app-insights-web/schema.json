{
    "$schema": "http://json-schema.org/schema",
    "cli": "nx",
    "$id": "AppInsightsWeb",
    "title": "Ensono Stacks Azure React App Insights Web Generator",
    "description": "The @ensono-stacks/azure-react:app-insights-web generator will install and configures a ReactJs app using the NX/React plugin then install and configures App Insights.",
    "type": "object",
    "properties": {
        "name": {
            "type": "string",
            "description": "Library name",
            "$default": {
                "$source": "argv",
                "index": 0
            },
            "x-prompt": "What name would you like to use for the library?",
            "pattern": "^[a-zA-Z].*$"
        },
        "connectionString": {
            "type": "string",
            "description": "The env variable for the connection string.",
            "x-prompt": "What is the env variable for the app insights connection string?"
        },
        "directory": {
            "type": "string",
            "description": "A directory where the lib is placed.",
            "alias": "dir"
        },
        "importPath": {
            "type": "string",
            "description": "What import path would you like to use for the library?"
        },
        "tags": {
            "type": "string",
            "description": "Add tags to the library (used for linting).",
            "alias": "t"
        }
    },
    "required": ["name", "connectionString"]
}
