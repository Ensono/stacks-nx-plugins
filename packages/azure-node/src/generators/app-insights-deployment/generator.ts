import {
    executedDependantGenerator,
    hasGeneratorExecutedForProject,
    verifyPluginCanBeInstalled,
} from '@ensono-stacks/core';
import { readProjectConfiguration, Tree } from '@nrwl/devkit';

import { AppInsightsDeploymentGeneratorSchema } from './schema';
import { updateDockerfile } from './utils/update-dockerfile';
import { updatePipelineStagesYaml } from './utils/update-pipeline';
import { updateProjectJsonHelmUpgradeTarget } from './utils/update-targets';

// eslint-disable-next-line consistent-return
export default async function appInsightsDeploymentGenerator(
    tree: Tree,
    options: AppInsightsDeploymentGeneratorSchema,
) {
    verifyPluginCanBeInstalled(tree, options.project);

    if (
        !executedDependantGenerator(
            tree,
            'WorkspaceDeployment',
            options.project,
        )
    )
        return false;
    if (
        hasGeneratorExecutedForProject(
            tree,
            options.project,
            'AzureNodeAppInsightsDeployment',
        )
    )
        return false;

    const project = readProjectConfiguration(tree, options.project);

    // Update project.json
    updateProjectJsonHelmUpgradeTarget(project, tree, options);
    updatePipelineStagesYaml(tree, options);
    updateDockerfile(project, tree);
}
