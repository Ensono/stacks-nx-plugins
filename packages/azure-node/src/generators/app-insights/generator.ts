import {
    tsMorphTree,
    addIgnoreEntry,
    formatFiles,
    thirdPartyDependencyWarning,
    deploymentGeneratorMessage,
    hasGeneratorExecutedForProject,
    verifyPluginCanBeInstalled,
} from '@ensono-stacks/core';
import {
    addDependenciesToPackageJson,
    getProjects,
    joinPathFragments,
    Tree,
} from '@nx/devkit';
import { runTasksInSerial } from '@nx/workspace/src/utilities/run-tasks-in-serial';
import chalk from 'chalk';
import path from 'path';

import { AppInsightsGeneratorSchema } from './schema';
import {
    initAppInsights,
    configureAppInsights,
    startAppInsights,
} from './templates/appInsights';
import { appInsightsVersion } from '../../../utils/versions';

function updateDependencies(tree: Tree) {
    return addDependenciesToPackageJson(
        tree,
        {
            applicationinsights: appInsightsVersion,
        },
        {},
    );
}

export default async function appInsightsGenerator(
    tree: Tree,
    options: AppInsightsGeneratorSchema,
) {
    verifyPluginCanBeInstalled(tree, options.project);

    if (
        hasGeneratorExecutedForProject(
            tree,
            options.project,
            'AzureNodeAppInsights',
        )
    )
        return false;

    const { applicationinsightsConnectionString, server } = options;
    const project = getProjects(tree).get(options.project);

    if (!project) {
        throw new Error(
            `No application was found with the name '${options.project}'`,
        );
    }

    const customServerPath = path.join(project.sourceRoot, server);

    // Check if custom server exist
    if (!tree.exists(customServerPath)) {
        throw new Error('No custom server found.');
    }

    const morphTree = tsMorphTree(tree);

    // Read the Node from the source file
    const customServer = morphTree.addSourceFileAtPath(customServerPath);

    const isAppInsightsImport = customServer
        .getImportDeclarations()
        .some(
            importDeclaration =>
                importDeclaration.getModuleSpecifier().getLiteralValue() ===
                'applicationinsights',
        );

    // Add appInsights to import if it's not there
    if (isAppInsightsImport) {
        throw new Error('AppInsights SDK is already in use.');
    }

    thirdPartyDependencyWarning(['@nx/next:custom-server']);

    customServer.addImportDeclaration({
        namespaceImport: 'appInsights',
        moduleSpecifier: 'applicationinsights',
    });

    const mainFunction = customServer.getFunction('main');

    // Add appInisghts statements
    mainFunction.insertStatements(
        0,
        initAppInsights(applicationinsightsConnectionString),
    );
    mainFunction.insertStatements(1, configureAppInsights(project.name));
    mainFunction.insertStatements(2, startAppInsights());

    // Add empty lines after appInisghts statements
    mainFunction.getStatements().forEach(statement => {
        if (statement.getText().includes('appInsights')) {
            statement.appendWhitespace(writer => writer.newLine());
        }
    });

    // Save changes
    customServer.saveSync();

    const serverPath = joinPathFragments(project.root, server);
    // add nrwl/next custom server to prettier ignore
    addIgnoreEntry(tree, '.prettierignore', 'next server', [serverPath]);
    // Format files excluding the server file
    await formatFiles(tree, [serverPath]);

    console.warn(
        chalk.yellow`${serverPath} has been added to .prettierignore; Amend this file to resolve linting issues.`,
    );

    // Add dependencies and install
    return runTasksInSerial(updateDependencies(tree), () =>
        deploymentGeneratorMessage(
            tree,
            `nx g @ensono-stacks/azure-node:app-insights-deployment --project ${options.project} --applicationinsightsConnectionString ${applicationinsightsConnectionString}`,
        ),
    );
}
