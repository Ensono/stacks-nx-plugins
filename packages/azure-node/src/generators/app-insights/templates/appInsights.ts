export const initAppInsights = (
    applicationinsightsConnectionString: string,
) => `appInsights
    .setup(process.env.${applicationinsightsConnectionString})
    .setAutoCollectConsole(true, true)
    .setAutoCollectDependencies(true)
    .setAutoCollectExceptions(true)
    .setAutoCollectHeartbeat(true)
    .setAutoCollectPerformance(true, true)
    .setAutoCollectRequests(true)
    .setAutoDependencyCorrelation(true)
    .setSendLiveMetrics(true)
    .setUseDiskRetryCaching(true);`;

export const configureAppInsights = (
    projectName: string,
) => `appInsights.defaultClient.context.tags[
    appInsights.defaultClient.context.keys.cloudRole
] = '${projectName}';`;

export const startAppInsights = () => `appInsights.start();`;
