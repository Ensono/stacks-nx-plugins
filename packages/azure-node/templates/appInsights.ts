export const initAppInsights = (appInsightsKey: string) => `appInsights
    .setup(process.env.${appInsightsKey})
    .setAutoCollectConsole(true, true)
    .setAutoCollectDependencies(true)
    .setAutoCollectExceptions(true)
    .setAutoCollectHeartbeat(true)
    .setAutoCollectPerformance(true, true)
    .setAutoCollectRequests(true)
    .setAutoDependencyCorrelation(true)
    .setSendLiveMetrics(true)
    .setUseDiskRetryCaching(true);`;

export const configureAppInsights = (
    projectName: string,
) => `appInsights.defaultClient.context.tags[
    appInsights.defaultClient.context.keys.cloudRole
] = '${projectName}';`;

export const startAppInsights = () => `appInsights.start();`;
