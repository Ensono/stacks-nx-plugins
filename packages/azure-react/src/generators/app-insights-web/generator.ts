import {
    NormalizedSchema as BaseNormalizedSchema,
    normalizeOptions,
} from '@ensono-stacks/core';
import {
    addDependenciesToPackageJson,
    formatFiles,
    generateFiles,
    names,
    offsetFromRoot,
    Tree,
} from '@nrwl/devkit';
import { Linter } from '@nrwl/linter';
import { libraryGenerator } from '@nrwl/react';
import path from 'path';

import {
    appInsightsReactVersion,
    appInsightsWebVersion,
} from '../../../utils/versions';
import { AppInsightsWebGeneratorSchema } from './schema';

type NormalizedSchema = BaseNormalizedSchema<AppInsightsWebGeneratorSchema>;

function updateDependencies(tree: Tree) {
    return addDependenciesToPackageJson(
        tree,
        {
            '@microsoft/applicationinsights-web': appInsightsWebVersion,
            '@microsoft/applicationinsights-react-js': appInsightsReactVersion,
        },
        {},
    );
}

function addFiles(tree: Tree, options: NormalizedSchema) {
    const templateOptions = {
        ...options,
        ...names(options.name),
        connectionString: names(options.connectionString).constantName,
        offsetFromRoot: offsetFromRoot(options.projectRoot),
        template: '',
    };
    generateFiles(
        tree,
        path.join(__dirname, 'files'),
        options.projectRoot,
        templateOptions,
    );
}

export default async function appInsightsWebGenerator(
    tree: Tree,
    options: AppInsightsWebGeneratorSchema,
) {
    if (!options.connectionString) {
        throw new Error('connectionString cannot be empty.');
    }

    // Normalize options
    const normalizedOptions = normalizeOptions(tree, options);

    // Use the existing library generator
    await libraryGenerator(tree, {
        ...options,
        linter: Linter.EsLint,
        style: 'none',
    });
    // Delete the default generated lib folder
    tree.delete(path.join(normalizedOptions.projectRoot, 'src', 'lib'));

    // Generate files
    addFiles(tree, normalizedOptions);

    // Format files
    await formatFiles(tree);

    // Update package.json and install dependencies
    return updateDependencies(tree);
}
