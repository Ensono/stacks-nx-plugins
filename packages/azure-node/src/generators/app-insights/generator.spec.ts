import { Tree, readProjectConfiguration, readJson } from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { applicationGenerator as nextGenerator } from '@nrwl/next/src/generators/application/application';

import * as appInisghtsTemplate from '../../../templates/appInsights';
import generator from './generator';
import { AppInsightsGeneratorSchema } from './schema';

describe('app-insights generator', () => {
    let appTree: Tree;
    const options: AppInsightsGeneratorSchema = {
        project: 'test',
        server: 'server.js',
        appInsightsKey: 'TEST_KEY',
    };

    beforeEach(() => {
        appTree = createTreeWithEmptyWorkspace();
        appTree.write(
            'test/server.js',
            `function main() { 
                //custom-server
            }`,
        );
    });

    it('should run successfully', async () => {
        await nextGenerator(appTree, {
            name: 'test',
            customServer: true,
            style: 'css',
        });

        const initAppInsightsSpy = jest.spyOn(
            appInisghtsTemplate,
            'initAppInsights',
        );
        const configureAppInsightsSpy = jest.spyOn(
            appInisghtsTemplate,
            'configureAppInsights',
        );
        const startAppInsightsSpy = jest.spyOn(
            appInisghtsTemplate,
            'startAppInsights',
        );

        await generator(appTree, options);

        expect(initAppInsightsSpy).toHaveBeenCalledWith('TEST_KEY');
        expect(configureAppInsightsSpy).toHaveBeenCalledWith('test');
        expect(startAppInsightsSpy).toHaveBeenCalled();
    });

    it('should throw an error when no application is found matching the project option', async () => {
        const incorrectOptions: AppInsightsGeneratorSchema = {
            project: 'unknown-project',
            server: 'server.js',
            appInsightsKey: 'TEST_KEY',
        };
        await expect(generator(appTree, incorrectOptions)).rejects.toThrowError(
            `No application was found with the name 'unknown-project'`,
        );
    });

    it('should throw an error when custom server is missing', async () => {
        await nextGenerator(appTree, {
            name: 'test',
            customServer: true,
            style: 'css',
        });

        appTree.delete('test/server.js');
        await expect(generator(appTree, options)).rejects.toThrowError(
            'No custom server found.',
        );
    });

    it('should throw an error if appinsights is already imported', async () => {
        await nextGenerator(appTree, {
            name: 'test',
            customServer: true,
            style: 'css',
        });

        appTree.write(
            'test/server.js',
            'import * as appInsights from "applicationinsights"',
        );
        await expect(generator(appTree, options)).rejects.toThrowError(
            'AppInsights SDK is already in use.',
        );
    });

    it('should install applicationinsights as dependency', async () => {
        await nextGenerator(appTree, {
            name: 'test',
            customServer: true,
            style: 'css',
        });

        await generator(appTree, options);

        const packageJson = readJson(appTree, 'package.json');
        expect(Object.keys(packageJson.dependencies)).toEqual(
            expect.arrayContaining(['applicationinsights']),
        );
    });
});
