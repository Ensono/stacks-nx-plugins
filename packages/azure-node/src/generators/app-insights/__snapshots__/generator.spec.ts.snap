// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app-insights generator executedGenerators should update server/main.ts 1`] = `
"/*
 * This is only a minimal custom server to get started.
 * You may want to consider using Express or another server framework, and enable security features such as CORS.
 *
 * For an example, see the Next.js repo:
 * Node - https://github.com/vercel/next.js/blob/canary/examples/custom-server
 */
import { createServer } from 'http';
import { parse } from 'node:url';
import * as path from 'path';
import next from 'next';
import * as appInsights from "applicationinsights";

// Next.js server options:
// - The environment variable is set by \`@nx/next:server\` when running the dev server.
// - The fallback \`__dirname\` is for production builds.
// - Feel free to change this to suit your needs.

const dir = process.env.NX_NEXT_DIR || path.join(__dirname, '..');
const dev = process.env.NODE_ENV === 'development';

// HTTP Server options:
// - Feel free to change this to suit your needs.
const hostname = process.env.HOST || 'localhost';
const port = process.env.PORT ? parseInt(process.env.PORT) : 4200;

async function main() {
    appInsights
        .setup(process.env.TEST_CONNECTION_STRING_ENV)
        .setAutoCollectConsole(true, true)
        .setAutoCollectDependencies(true)
        .setAutoCollectExceptions(true)
        .setAutoCollectHeartbeat(true)
        .setAutoCollectPerformance(true, true)
        .setAutoCollectRequests(true)
        .setAutoDependencyCorrelation(true)
        .setSendLiveMetrics(true)
        .setUseDiskRetryCaching(true);

    appInsights.defaultClient.context.tags[
        appInsights.defaultClient.context.keys.cloudRole
    ] = 'test';

    appInsights.start();

  const nextApp = next({ dev, dir });
  const handle = nextApp.getRequestHandler();

  await nextApp.prepare();

  const server = createServer((req, res) => {
    const parsedUrl = parse(req.url ?? '', true);
    handle(req, res, parsedUrl);
  });

  server.listen(port, hostname);

  console.log(\`[ ready ] on http://\${hostname}:\${port}\`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
"
`;
